File: frontend/src/types/issue.ts
/**
 * TypeScript types mirroring /gim/src/models/issue.py
 */

export type IssueStatus = "active" | "superseded" | "invalid";

export type ContributionType = "environment" | "symptom" | "model_quirk" | "validation";

export type RootCauseCategory =
  | "environment"
  | "model_behavior"
  | "api_integration"
  | "code_generation"
  | "framework_specific";

export interface MasterIssue {
  id: string;
  canonical_title: string;
  description: string;
  root_cause_category: RootCauseCategory;
  root_cause_subcategory?: string;
  confidence_score: number;
  child_issue_count: number;
  environment_coverage: string[];
  verification_count: number;
  last_confirmed_at?: string;
  status: IssueStatus;
  created_at: string;
  updated_at: string;
  model_provider?: string;
  language?: string;
  framework?: string;
}

export interface MasterIssueCreate {
  canonical_title: string;
  description: string;
  root_cause_category: RootCauseCategory;
  root_cause_subcategory?: string;
}

export interface ChildIssueCreate {
  master_issue_id: string;
  contribution_type: ContributionType;
  sanitized_error: string;
  sanitized_context?: string;
  sanitized_mre?: string;
  environment: Record<string, unknown>;
  model_provider: string;
  model_name: string;
  model_version?: string;
  model_behavior_notes: string[];
  validation_success?: boolean;
  validation_notes?: string;
}

export interface ChildIssue extends ChildIssueCreate {
  id: string;
  created_at: string;
}

/**
 * Child issue detail returned by GET /issues/{childId} when the ID is a child.
 */
export interface ChildIssueDetail {
  id: string;
  is_child_issue: true;
  master_issue_id: string;
  original_error: string;
  original_context: string;
  code_snippet: string;
  model: string;
  provider: string;
  language: string;
  framework: string;
  submitted_at: string;
  contribution_type: string;
  validation_success: boolean | null;
  metadata: Record<string, unknown>;
  parent_canonical_title?: string;
  parent_root_cause_category?: RootCauseCategory;
}

/**
 * Child issue list item returned by GET /issues/{id}/children.
 */
export interface ChildIssueListItem {
  id: string;
  master_issue_id: string;
  original_error: string;
  provider: string;
  language: string;
  framework: string;
  submitted_at: string;
  contribution_type: string;
  model_name: string;
  validation_success: boolean | null;
}

/**
 * Response from GET /issues/{id}/children.
 */
export interface ChildIssueListResponse {
  children: ChildIssueListItem[];
  total: number;
  limit: number;
  offset: number;
  master_issue_id: string;
}

/**
 * Category display configuration
 */
export const CATEGORY_DISPLAY: Record<RootCauseCategory, { label: string; color: string }> = {
  environment: { label: "Environment", color: "cat-environment" },
  model_behavior: { label: "Model", color: "cat-model" },
  api_integration: { label: "API", color: "cat-api" },
  code_generation: { label: "Codegen", color: "cat-codegen" },
  framework_specific: { label: "Framework", color: "cat-framework" },
};

/**
 * Status display configuration
 */
export const STATUS_DISPLAY: Record<IssueStatus, { label: string; variant: "success" | "warning" | "error" }> = {
  active: { label: "Verified", variant: "success" },
  superseded: { label: "Pending", variant: "warning" },
  invalid: { label: "Declined", variant: "error" },
};

File: gim/src/models/issue.py
"""Pydantic models for Master and Child issues."""

from datetime import datetime
from enum import Enum
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class IssueStatus(str, Enum):
    """Status of a master issue.

    Attributes:
        ACTIVE: Issue is currently active and valid.
        SUPERSEDED: Issue has been replaced by a newer issue.
        INVALID: Issue has been marked as invalid.
    """

    ACTIVE = "active"
    SUPERSEDED = "superseded"
    INVALID = "invalid"


class ContributionType(str, Enum):
    """Type of contribution a child issue makes.

    Attributes:
        ENVIRONMENT: Adds new environment information.
        SYMPTOM: Adds new symptoms or error variations.
        MODEL_QUIRK: Documents model-specific behavior.
        VALIDATION: Confirms or validates the fix.
    """

    ENVIRONMENT = "environment"
    SYMPTOM = "symptom"
    MODEL_QUIRK = "model_quirk"
    VALIDATION = "validation"


class RootCauseCategory(str, Enum):
    """Root cause classification categories.

    Attributes:
        ENVIRONMENT: Environment-related issues (dependencies, config).
        MODEL_BEHAVIOR: AI model behavior issues (tool calling, schema).
        API_INTEGRATION: API/integration issues (auth, network).
        CODE_GENERATION: Code generation issues (syntax, types).
        FRAMEWORK_SPECIFIC: Framework-specific issues.
    """

    ENVIRONMENT = "environment"
    MODEL_BEHAVIOR = "model_behavior"
    API_INTEGRATION = "api_integration"
    CODE_GENERATION = "code_generation"
    FRAMEWORK_SPECIFIC = "framework_specific"


class MasterIssueBase(BaseModel):
    """Base model for master issue data.

    Attributes:
        canonical_title: Searchable title for the issue.
        description: Detailed description of the issue.
        root_cause_category: Primary root cause classification.
        root_cause_subcategory: Specific subcategory within root cause.
    """

    model_config = ConfigDict(str_strip_whitespace=True)

    canonical_title: str = Field(
        ...,
        min_length=10,
        max_length=500,
        description="Canonical, searchable title for the issue",
    )
    description: str = Field(
        ...,
        min_length=20,
        description="Detailed description of the issue",
    )
    root_cause_category: RootCauseCategory = Field(
        ...,
        description="Primary root cause classification",
    )
    root_cause_subcategory: Optional[str] = Field(
        None,
        max_length=100,
        description="Specific subcategory within root cause",
    )


class MasterIssueCreate(MasterIssueBase):
    """Model for creating a new master issue.

    Inherits all fields from MasterIssueBase.
    """

    pass


class MasterIssueResponse(MasterIssueBase):
    """Model for master issue API responses.

    Attributes:
        id: Unique identifier for the issue.
        confidence_score: Trust score between 0.0 and 1.0.
        child_issue_count: Number of child issues.
        environment_coverage: List of environments where fix is confirmed.
        verification_count: Number of successful fix verifications.
        last_confirmed_at: Timestamp of last successful confirmation.
        status: Current status of the issue.
        created_at: Timestamp when issue was created.
        updated_at: Timestamp when issue was last updated.
        source: Origin of the issue ('mcp_tool' or 'github_crawler').
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    confidence_score: float = Field(ge=0.0, le=1.0)
    child_issue_count: int = Field(ge=0)
    environment_coverage: List[str] = Field(default_factory=list)
    verification_count: int = Field(ge=0)
    last_confirmed_at: Optional[datetime] = None
    status: IssueStatus
    created_at: datetime
    updated_at: datetime
    source: Optional[str] = None


class ChildIssueCreate(BaseModel):
    """Model for creating a child issue contribution.

    Attributes:
        master_issue_id: ID of the parent master issue.
        contribution_type: Type of contribution being made.
        sanitized_error: Sanitized error message.
        sanitized_context: Sanitized context around the error.
        sanitized_mre: Sanitized minimal reproducible example.
        environment: Environment information as dict.
        model_provider: Provider of the AI model.
        model_name: Name of the AI model.
        model_version: Version of the AI model.
        model_behavior_notes: Notes about model-specific behavior.
        validation_success: Whether the fix was validated successfully.
        validation_notes: Notes about the validation.
    """

    model_config = ConfigDict(str_strip_whitespace=True)

    master_issue_id: UUID
    contribution_type: ContributionType
    sanitized_error: str = Field(..., min_length=5)
    sanitized_context: Optional[str] = None
    sanitized_mre: Optional[str] = None
    environment: dict = Field(default_factory=dict)
    model_provider: str = Field(..., min_length=1, max_length=50)
    model_name: str = Field(..., min_length=1, max_length=100)
    model_version: Optional[str] = None
    model_behavior_notes: List[str] = Field(default_factory=list)
    validation_success: Optional[bool] = None
    validation_notes: Optional[str] = None


class ChildIssueResponse(ChildIssueCreate):
    """Model for child issue API responses.

    Attributes:
        id: Unique identifier for the child issue.
        created_at: Timestamp when child issue was created.
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    created_at: datetime
